### ISSUE TITLE
Image cleaning not working as expected when tags get overwritten

### ISSUE THREAD

**Issue by skorpios** [aws/amazon-ecs-agent#1126]
### Summary
The ECS agent seems to not clean images with no tag, or when a different tag of the same repository is currently in use by a container.

### Description
We have a testing environment built on ECS and Github webhooks. For every push in a Pull Request, a new image is pushed to ECR using the PR number as a tag. Then the image is deployed to an ECS cluster. The image name looks like this:
```
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/application_name:pr_1024
```
Each time a PR is updated, a different image will be pushed to the same tag, meaning that previous images with that tag will now have no tag.

At some point we noticed that we were running out of space, and that some images were not cleaned by the ECS agent. Here is a list of images on one instance in the cluster:
```
REPOSITORY                                                         TAG                 IMAGE ID            CREATED             SIZE
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app2         pr_593              6c5fac090c92        2 hours ago         1.15GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app2         <none>              cb4c1069f759        2 hours ago         1.15GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_master           0bbbd3898a4d        2 hours ago         1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10295            81a678cf5539        5 hours ago         1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10297            68a7d0e8da94        7 hours ago         1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10296            6e5d523a22c6        2 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         <none>              1682fc144a2c        2 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         <none>              85377b421018        3 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10294            a78736a6233a        3 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         <none>              af3dd2091a0e        3 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10293            1f8ae42a98ea        3 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10291            a3d02aa492dd        3 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         <none>              969593d65952        3 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10290            08e1ce9e8cad        3 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10289            80ba8fa2233d        3 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         <none>              597807c3df13        4 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         pr_10287            9f684f92d5c7        4 days ago          1.64GB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app3         <none>              d47a9cb9db03        4 days ago          632MB
xxxxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/previews/app1         <none>              32cd674a0849        4 days ago          1.64GB
```

Because the applications are meant to be short-lived, the ECS agent is configured to clean containers and images pretty aggressively.

### Expected Behavior
I was expecting most of the images in the list above to be removed by the agent.

I suspect the `previews/app3:none` image not being removed because it has no tag.

For the other applications, there will always be at least 1 task running 1 tag of the image, but I would assume the tags that are not used by any container would be removed.

### Observed Behavior
Some images are still kept although there is no running or stopped container that relies on them.

### Environment Details
Content of `/etc/ecs/ecs.config`:
```
ECS_ENGINE_AUTH_TYPE=dockercfg
ECS_ENGINE_AUTH_DATA=REDACTED
ECS_ENGINE_TASK_CLEANUP_WAIT_DURATION=1m
ECS_IMAGE_CLEANUP_INTERVAL=10m
ECS_IMAGE_MINIMUM_CLEANUP_AGE=10m
ECS_NUM_IMAGES_DELETE_PER_CYCLE=100
ECS_CLUSTER=previews-201711
```
The instances are running the ECS-optimized AMI with very little customization.

Output of `docker info`:
```
Containers: 5
 Running: 5
 Paused: 0
 Stopped: 0
Images: 24
Server Version: 17.06.2-ce
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: ext4
 Data file:
 Metadata file:
 Data Space Used: 12.01GB
 Data Space Total: 23.33GB
 Data Space Available: 11.32GB
 Metadata Space Used: 6.607MB
 Metadata Space Total: 25.17MB
 Metadata Space Available: 18.56MB
 Thin Pool Minimum Free Space: 2.333GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: fluentd
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.62-21.56.amzn1.x86_64
Operating System: Amazon Linux AMI 2017.09
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.798GiB
Name: ip-192-168-113-107
ID: FI25:7IOZ:N5K3:55XW:CMBR:O3HL:CLWI:AS4H:ZNXM:GLGX:XHJM:BTRH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

### Supporting Log Snippets
This is what gets logged on every cleanup cycle:
```
2017-12-04T16:01:19Z [INFO] Begin building map of eligible unused images for deletion
2017-12-04T16:01:19Z [INFO] No eligible images for deletion for this cleanup cycle
2017-12-04T16:01:19Z [INFO] End of eligible images for deletion: No more eligible images for deletion; Still have 2 image states being managed
```
It mentions `2 image states being managed` although there should be more.

**Comment by richardpen** [comment]
@skorpios Sorry for the late response, based the description above it looks like you're running into the same issue as #558 which we have fixed since ECS Agent v1.13.1. Are you running the latest agent? If you are still experiencing this issue, could you send me the logs along with the images that are not used by any containers? You can collect the logs with [Amazon ECS Log Collector](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-logs-collector.html) and send it to me: penyin (at) amazon.com

Thanks,
Peng

**Comment by richardpen** [comment]
@skorpios Since we haven't heard from you for a long time, I'm closing this issue for now. Feel free to reopen this and provide the information requested above if you run into this issue in the future.
